### PR TITLE
Added OLED "rewiring" menu, added corresponding images/fonts necessary 

### PR DESCRIPTION
**Pull request by TABYDACAT**
- OLED menu added for basic selection of features
- Menu includes: Resolution selection, preset slots, resetting GBS and loading defaults 
- Separate menu selection to display info for the following: Input type, resolutions and frame rate
- Menu controlled via Rotary Encoder which is done with ISR 
- Consoles tested: PS1,PS2, and DC (RGB, Component, VGA) 

Here's a link to short video of menu working
https://twitter.com/i/status/1395823042228535300

### CODE REVIEW

**Comment by gingerbeardman** [comment]
Very cool.

Do you have info about how to wire up an OLED?

**Review by ramapcsx2 — COMMENTED** [review]
Hey, thanks for this PR!
I'll add a few requests for changes :)

File: gbs-control.ino [hunk 1]
```diff
@@ -20,6 +20,30 @@
...
+int selectOption = 0;
+int page = 0;
+
+String menu[4] = {"Resolutions","Presets","Misc.","Current Settings"};
+String Resolutions[7] = {"1280x960","1280x1024","1280x720","1920x1080","480/576","Downscale","Pass-Through"};
+String Presets[8] = {"1","2","3","4","5","6","7","Back"};
+String Misc[4] = {"Reset GBS","Restore Factory","-----Back"};
+
+int lastCount = 0;
+volatile int encoder_pos = 0;
```
Comment (ramapcsx2) [inline gbs-control.ino:42]: Can these OLED related variables get an "OLED" attachment? This is to clearly separate them from GBSC base code.

File: gbs-control.ino [hunk 2]
```diff
@@ -6917,8 +6941,39 @@ void loadDefaultUserOptions() {
 //  //system_phy_set_powerup_option(3); // 0 = default, use init byte; 3 = full calibr. each boot, extra 200ms
 //  system_phy_set_powerup_option(0);
 //}
-
+void ICACHE_RAM_ATTR isr (){
```
Comment (ramapcsx2) [inline gbs-control.ino:6944]: isrOLED ?

File: gbs-control.ino [hunk 3]
```diff
@@ -6917,8 +6941,39 @@ void loadDefaultUserOptions() {
...
+      sub_pointer-=15;
+      pointer_count--;
+      // down = false;
+      // up = true;
+    }
+  }
+  lastInterruptTime = inturruptTime;
+}
 void setup() {
+  display.init();
```
Comment (ramapcsx2) [inline gbs-control.ino:6968]: displayOLED?

File: gbs-control.ino [hunk 4]
```diff
@@ -7415,7 +7472,12 @@ void loop() {
   static unsigned long lastTimeSyncWatcher = millis();
   static unsigned long lastTimeSourceCheck = 500; // 500 to start right away (after setup it will be 2790ms when we get here)
   static unsigned long lastTimeInterruptClear = millis();
-
+  
+  settingsMenu();
```
Comment (ramapcsx2) [inline gbs-control.ino:7476]: Here it's very obvious how the "settingsMenu" could be confused for a GBSC funtion :)

**Review by ramapcsx2 — CHANGES_REQUESTED** [review]
Looks like only a few changes required.
We'll need to know which display to go buy, and how to wire it up! :)

**Comment by xperiments** [comment]
Hi, First of all thanks for the code, finally you uploaded it since we speak last time XD.

Checking quickly your code I see some areas of refactoring, mostly naming variables / structures and also I think we can simplify a lot of repeating code. 

The Display thing needs his own file, aside from the main .ino code.

Also I think to better provide a general "display" api that then can be used by the SSD1306 "driver". This way it can be replaced by other displays "drivers" as needed.

Good work BTW!!!

**Comment by ramapcsx2** [comment]
@xperiments: @TABYDACAT is more of an electrical engineer, not purely a coder. I think it'll be fine with just the variable name clarifications. The API thing is a bit much to ask :)

**Comment by xperiments** [comment]
@ramapcsx2 Of course, was with the intention to help XD, already contacted him by twitter XD. Anyway some changes will be needed... at least a #define to define the presence of the OLED etc... So people does not have the oled + pot can compile without this dependencies...

File: gbs-control.ino (hunk 2 above)
Comment (xperiments) [reply gbs-control.ino:6944]: Not this is for the Rotatory pot I think

Comment (TABYDACAT) [reply gbs-control.ino:6944]: This the Interrupt Service Routine for the rotary encoder. I opted for this method in hopes that it wouldn't slow the rest of the program. Typically, I was told that an ISR goes before void setup, but can fix placement if necessary

File: gbs-control.ino (hunk 1 above)
Comment (TABYDACAT) [reply gbs-control.ino:42]: Just simply adding on an OLED tag to them? Sure! For example, "volatile int oled_encoder_pos = 0"?

File: gbs-control.ino (hunk 4 above)
Comment (TABYDACAT) [reply gbs-control.ino:7476]: Probably will need OLED rename to show its part of that instead

File: gbs-control.ino (hunk 2 above)
Comment (ramapcsx2) [reply gbs-control.ino:6944]: "isrRotaryEncoder" is good then. The naming is just to be instantly in the know what the ISR is for when browsing the code :)

File: gbs-control.ino (hunk 4 above)
Comment (ramapcsx2) [reply gbs-control.ino]: settingsMenuOLED(); then yep :)

File: gbs-control.ino (hunk 1 above)
Comment (ramapcsx2) [reply gbs-control.ino:42]: yep, sounds good

**Comment by webhdx** [comment]
I love the idea. As far as the code goes I think it would much needed to move it to a separate file. `gbs-control.ino` is now very difficult to manage since there is almost everything packed in there.

File: gbs-control.ino (hunk 2 above)
Comment (TABYDACAT) [reply gbs-control.ino]: edited to reflect this change

File: gbs-control.ino (hunk 4 above)
Comment (TABYDACAT) [reply gbs-control.ino]: edited to reflect this change

**Comment by TABYDACAT** [comment]
@webhdx Hey thnx for trying out the code! It’s actually the first time I’ve seen it on someone else’s setup! 

As for the suggestions, I tried exactly with what you’re referring to with the  `current settings `. I had it sort-of working with a “hold button” check. Holding the button for a few seconds, increment a container, store and compare to check what screen you were currently on. This resulted in weird effects whether it being the ESP crashing or screen flickering. I’ll work on this some more in the meantime though. I have some options to try out

File: gbs-control.ino (hunk 3 above)
Comment (ramapcsx2) [reply gbs-control.ino:6968]: Ah okay, then a comment will do. A non-existing OLED display device on the I2C bus will simply not respond, and the code move on, right?

Comment (TABYDACAT) [reply gbs-control.ino:6968]: Yes, if no OLED device is connected on the bus, it will compile fine and operate as normal :)

**Comment by ramapcsx2** [comment]
@webhdx:
Could you quickly confirm a few things for me?
- What happens if the extra library for the display isn't installed? Compilation fails, right?
- What happens if you remove the display I2C cables? Does the software work normally otherwise?

I'm a bit swamped in xStation work (tricky stuff! :p), so it'd be great to hear about these 2 points.
If they're good to go, I can add the required library section to the wiki, and we can go ahead and merge this.
Future improvements can be made easily. They don't all have to be in this PR.

Cheers :)

File: gbs-control.ino (hunk 3 above)
Comment (TABYDACAT) [reply gbs-control.ino]: Comments added. Edited to reflect this change

**Comment by TABYDACAT** [comment]
> @webhdx:
> Could you quickly confirm a few things for me?
> 
> * What happens if the extra library for the display isn't installed? Compilation fails, right?
> * What happens if you remove the display I2C cables? Does the software work normally otherwise?
> 
> I'm a bit swamped in xStation work (tricky stuff! :p), so it'd be great to hear about these 2 points.
> If they're good to go, I can add the required library section to the wiki, and we can go ahead and merge this.
> Future improvements can be made easily. They don't all have to be in this PR.
> 
> Cheers :)

Just wanted to chime in here if that's okay. Yes, compilation will fail if the Thingpulse library isn't installed. Luckily, it's part of the Arduino library manager so no need to do it manually.

If you don't have the OLED connected the software will operate normally. Currently, I have it setup so that the D1 mini and OLED can be removed from circuit at any time. Through my test of removing the OLED/Rotary Encoder from circuit the GBS-C software acts as normal not affecting normal usage/gameplay whatsoever.

However, yes I'm interested to see what others testing comes about with this. Perhaps I missed something so all ears :)

**Comment by ramapcsx2** [comment]
There seems to be some kind of conflict, but I can't parse what git is trying to tell me what the conflict is :p

**Comment by TABYDACAT** [comment]
> There seems to be some kind of conflict, but I can't parse what git is trying to tell me what the conflict is :p

Did it update with no conflicts now? It's showing no conflicts on my end

**Review by ramapcsx2 — APPROVED** [review]

**Comment by ramapcsx2** [comment]
Does anyone happen to know whether the whitespace changes can be undone or ignored in the merge somehow?
https://i.imgur.com/QeHezVz.png
If not, the entire gbscontrol.ino will show up as changed.
It's not the end of the world though :p

**Comment by TABYDACAT** [comment]
> Does anyone happen to know whether the whitespace changes can be undone or ignored in the merge somehow?
> https://i.imgur.com/QeHezVz.png
> If not, the entire gbscontrol.ino will show up as changed.
> It's not the end of the world though :p

Don't know of a way, sorry! Perhaps starting a new pull request, and i'll just push the fork I have, just editing the OLED stuff?

**Comment by xperiments** [comment]
> Does anyone happen to know whether the whitespace changes can be undone or ignored in the merge somehow?
> https://i.imgur.com/QeHezVz.png
> If not, the entire gbscontrol.ino will show up as changed.
> It's not the end of the world though :p

Yeah, if you use VSCode you can save changes without formating with CMD+R+S in mac, win CRTL+R+S ? not sure...

This way code is not changed. Is the way I have used to introduce as little changes as possible

**Comment by TABYDACAT** [comment]
> > Does anyone happen to know whether the whitespace changes can be undone or ignored in the merge somehow?
> > https://i.imgur.com/QeHezVz.png
> > If not, the entire gbscontrol.ino will show up as changed.
> > It's not the end of the world though :p
> 
> Yeah, if you use VSCode you can save changes without formating with CMD+R+S in mac, win CRTL+R+S ? not sure...
> 
> This way code is not changed. Is the way I have used to introduce as little changes as possible.
> 
> So one way is copy the changes aside, clone master, create a branch from it, add this changes using the CRTL thing.

Ah okay! Should I do this step?

**Comment by xperiments** [comment]
> > > Does anyone happen to know whether the whitespace changes can be undone or ignored in the merge somehow?
> > > https://i.imgur.com/QeHezVz.png
> > > If not, the entire gbscontrol.ino will show up as changed.
> > > It's not the end of the world though :p
> > 
> > 
> > Yeah, if you use VSCode you can save changes without formating with CMD+R+S in mac, win CRTL+R+S ? not sure...
> > This way code is not changed. Is the way I have used to introduce as little changes as possible.
> > So one way is copy the changes aside, clone master, create a branch from it, add this changes using the CRTL thing.
> 
> Ah okay! Should I do this step?

Up to you, I think we can go as is now, as actually the space thing will be removed and "cleaned" for future commits.

IDK if there is a common .ino or c+ code formatter, if so, maybe we can use it in the future

But for now yeah the CRTL thing makes sense

**Comment by TABYDACAT** [comment]
> > > > Does anyone happen to know whether the whitespace changes can be undone or ignored in the merge somehow?
> > > > https://i.imgur.com/QeHezVz.png
> > > > If not, the entire gbscontrol.ino will show up as changed.
> > > > It's not the end of the world though :p
> > > 
> > > 
> > > Yeah, if you use VSCode you can save changes without formating with CMD+R+S in mac, win CRTL+R+S ? not sure...
> > > This way code is not changed. Is the way I have used to introduce as little changes as possible.
> > > So one way is copy the changes aside, clone master, create a branch from it, add this changes using the CRTL thing.
> > 
> > 
> > Ah okay! Should I do this step?
> 
> Up to you, I think we can go as is now, as actually the space thing will be removed and "cleaned" for future commits.
> 
> IDK if there is a common .ino or c+ code formatter, if so, maybe we can use it in the future
> 
> But for now yeah the CRTL thing makes sense

Ah okay if it isn’t going to cause problems later I’d say go as is (w/ CTRL) but just my opinion

**Pull request closed without merging**